pkg/build/controller/controller.go

```diff
@@ -51,39 +58,77 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) {
 		return
 	}
 
-	nextStatus := buildapi.BuildStatusFailed
-	build.PodName = fmt.Sprintf("build-%s", build.Name)
+	if err := bc.nextBuildStatus(build); err != nil {
+		glog.V(4).Infof("Build failed with error %s/%s: %#v", build.Namespace, build.Name, err)
+		build.Status = buildapi.BuildStatusError
+		build.Message = err.Error()
+	}
 
+	if _, err := bc.BuildUpdater.UpdateBuild(build.Namespace, build); err != nil {
+		glog.V(2).Infof("Failed to record changes to build %s/%s: %#v", build.Namespace, build.Name, err)
+	}
+}
+
+// nextBuildStatus updates build with any appropriate changes, or returns an error if
+// the change cannot occur. When returning nil, be sure to set build.Status and optionally
+// build.Message.
+func (bc *BuildController) nextBuildStatus(build *buildapi.Build) error {
 	// If a cancelling event was triggered for the build, update build status.
 	if build.Cancelled {
-		glog.V(2).Infof("Cancelling build %s.", build.Name)
-		nextStatus = buildapi.BuildStatusCancelled
-
-	} else {
-
-		var podSpec *kapi.Pod
-		var err error
-		if podSpec, err = bc.BuildStrategy.CreateBuildPod(build); err != nil {
-			glog.V(2).Infof("Strategy failed to create build pod definition: %v", err)
-			nextStatus = buildapi.BuildStatusFailed
-		} else {
-
-			if _, err := bc.PodManager.CreatePod(build.Namespace, podSpec); err != nil {
-				if !errors.IsAlreadyExists(err) {
-					glog.V(2).Infof("Failed to create pod for build %s: %#v", build.Name, err)
-					nextStatus = buildapi.BuildStatusFailed
-				}
-			} else {
-				glog.V(2).Infof("Created build pod: %#v", podSpec)
-				nextStatus = buildapi.BuildStatusPending
+		glog.V(4).Infof("Cancelling build %s.", build.Name)
+		build.Status = buildapi.BuildStatusCancelled
+		return nil
+	}
+
+	// lookup the destination from the referenced image repository
+	spec := build.Parameters.Output.DockerImageReference
```

Comment:
Late binding - you can create builds individually, and the references should be resolved at bind time.  Remember, the build could be created before the image repository is created.  Eventually, this needs to put the build back into an error state and it should retry creating the pod at some point.

----- Original Message -----
> > -
> > -			if _, err := bc.PodManager.CreatePod(build.Namespace, podSpec); err !=
> > nil {
> > -				if !errors.IsAlreadyExists(err) {
> > -					glog.V(2).Infof("Failed to create pod for build %s: %#v", build.Name,
> > err)
> > -					nextStatus = buildapi.BuildStatusFailed
> > -				}
> > -			} else {
> > -				glog.V(2).Infof("Created build pod: %#v", podSpec)
> > -				nextStatus = buildapi.BuildStatusPending
> > +		glog.V(4).Infof("Cancelling build %s.", build.Name)
> > +		build.Status = buildapi.BuildStatusCancelled
> > +		return nil
> > +	}
> > +
> > +	// lookup the destination from the referenced image repository
> > +	spec := build.Parameters.Output.DockerImageReference
> 
> this functionality doesn't seem like it belongs in a method named
> "nextBuildStatus".  Why didn't you setup the DockerImageReference when the
> build was created?
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/679/files#r23220914